### PR TITLE
[iMarari] Added Generics task

### PR DESCRIPTION
**Pull request by iMarari**

### CODE REVIEW

File: iuriMorari/src/main/java/Generics/AllMedia.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+                '}';
+    }
+
+    public String getCategory() {
+        return category;
+    }
+
+    public Double getCost() {
+        if (cost >= 0)
+            this.cost = cost;
```
Comment (asecu) [inline iuriMorari/src/main/java/Generics/AllMedia.java:31]: your method is used in order to return the value of Media's cost. However, you have defined an setter, that doesn't have any value as method parameter. It should look like :
public Double getCost() {
return cost;
}

File: iuriMorari/src/main/java/Generics/AllMedia.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        else System.out.println("Negative cost of ittem. It need to be positive or 0");
+        return cost;
+    }
+
+    public void setCost(Double cost) {
+        if (cost >= 0)
+            this.cost = cost;
+        else System.out.println("Negative cost of ittem. It need to be positive or 0");
+    }
```
Comment (asecu) [inline iuriMorari/src/main/java/Generics/AllMedia.java:40]: can we add some methods in order to get access control for all Media's variables? (Getters and Setters)

File: iuriMorari/src/main/java/Generics/Books.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    private int pagesNumber;
+    private Boolean pictures;
+
+    public Books(String name, String category, Double cost, String ageName, int pagesNumber, Boolean pictures) {
+        super(name, category, cost);
+        this.ageName = ageName;
+        this.pagesNumber = pagesNumber;
+        this.pictures = pictures;
+    }
+
```
Comment (asecu) [inline iuriMorari/src/main/java/Generics/Books.java:14]: all the variables are private, the access is limited? no manipulation with data? Can we define at least 2 methods in order to use the Book's variables

File: iuriMorari/src/main/java/Generics/GenericLibrary.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    public void addItems(T t) {
+        allList.add(t);
+    }
+
+    public void deleteItems(T t) {
+        allList.remove(t);
+    }
+
+    public void showNames(T t) {
+        System.out.println(t);
```
Comment (asecu) [inline iuriMorari/src/main/java/Generics/GenericLibrary.java:23]: there is necessary to adjust this method. Here is called the toString() method. Could you please review the implementation of showNames() method, in order to print only the name of the media.

File: iuriMorari/src/main/java/Generics/GenericLibrary.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    public void filterByCost() {
+        allList.sort(Comparator.comparing(T::getCost));
```
Comment (asecu) [inline iuriMorari/src/main/java/Generics/GenericLibrary.java:27]: filter != sort -> By filtering we mean that we will get only records or data that belongs to some criteria, there is not that moment. Please, review the implementation of the method in order to get a filter method.

File: iuriMorari/src/main/java/Generics/GenericLibrary.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    public void filterByCategory() {
+        allList.sort(Comparator.comparing(T::getCategory));
```
Comment (asecu) [inline iuriMorari/src/main/java/Generics/GenericLibrary.java:31]: filterByCategory() means to get the data which belongs to a specific category, let's introduce a method parameter which is going to be the filter option, then make sure that you go through the list and return only methods that belongs to this category.

File: iuriMorari/src/main/java/Generics/GenericLibrary.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+    public List<T> getAllList() {
+        return allList;
+    }
+
+    public void addItems(T t) {
+        allList.add(t);
+    }
+
+    public void deleteItems(T t) {
```
Comment (asecu) [inline iuriMorari/src/main/java/Generics/GenericLibrary.java:18]: deleteItems() or deleteItem(), let's use correct syntax in order to provide quality code

File: iuriMorari/src/main/java/Generics/GenericLibrary.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+import java.util.List;
+
+public class GenericLibrary<T extends AllMedia> {
+    private List<T> allList = new ArrayList<>();
+
+    public List<T> getAllList() {
+        return allList;
+    }
+
+    public void addItems(T t) {
```
Comment (asecu) [inline iuriMorari/src/main/java/Generics/GenericLibrary.java:14]: addItems() or addItem(), let's use appropriate names

File: iuriMorari/src/main/java/Generics/GenericLibrary.java [hunk 9]
```diff
@@ -0,0 +1,41 @@
+package Generics;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+
+public class GenericLibrary<T extends AllMedia> {
+    private List<T> allList = new ArrayList<>();
```
Comment (asecu) [inline iuriMorari/src/main/java/Generics/GenericLibrary.java:8]: allList is not a suggestive name, how about mediaList or resourcesList?

File: iuriMorari/src/main/java/Generics/Library.java [hunk 10]
```diff
@@ -0,0 +1,21 @@
+package Generics;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Library {
+    private List<AllMedia> allMedia;
+
+    public Library() {
+        allMedia = new ArrayList<>();
```
Comment (asecu) [inline iuriMorari/src/main/java/Generics/Library.java:10]: can we set the initialization out of the Constructor?

File: iuriMorari/src/main/java/Generics/Newspaper.java [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+    public Newspaper(String name, String category, Double cost, String kindOfPaper, String author, int numberOfCopies) {
+        super(name, category, cost);
+        this.kindOfPaper = kindOfPaper;
+        this.author = author;
+        this.numberOfCopies = numberOfCopies;
+    }
+
+    @Override
+    public String toString() {
```
Comment (asecu) [inline iuriMorari/src/main/java/Generics/Newspaper.java:16]: by calling this toString() method will be difficult (impossible) to see the newspaper name, which is quite strange. Does we need the paper type, author and number of copies or Name is also important ?

File: iuriMorari/src/main/java/Generics/PrintClass.java [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+        for (T t1 : t) {
+            System.out.print(t1.toString() + " ");
+        }
+        System.out.println();
+        for (S s1 : s) {
+            System.out.println(s1.toString() + " ");
+        }
+    }
+
+    public static <T> void printArray(T[] generic) {
```
Comment (asecu) [inline iuriMorari/src/main/java/Generics/PrintClass.java:16]: use appropriate variable names generics -> objectArray

File: iuriMorari/src/main/java/Generics/Video.java [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+    public Video(String name, String category, Double cost, String kindOfMovie, String storageCarrier, int duration) {
+        super(name, category, cost);
+        this.kindOfMovie = kindOfMovie;
+        this.storageCarrier = storageCarrier;
+        this.duration = duration;
+    }
+
+    @Override
+    public String toString() {
```
Comment (asecu) [inline iuriMorari/src/main/java/Generics/Video.java:16]: alt + insert is not a solution, it's a short cut, please refactor it. Add the main details about Video in toString() method

File: iuriMorari/src/main/java/Generics/Video.java [hunk 14]
```diff
@@ -0,0 +1,23 @@
+package Generics;
+
+public class Video extends AllMedia {
+    private String kindOfMovie;
+    private String storageCarrier;
+    private int duration;
```
Comment (asecu) [inline iuriMorari/src/main/java/Generics/Video.java:6]: can we declare this variable as Integer? What is the difference?

File: iuriMorari/src/main/java/Generics/PrintClass.java [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+import java.util.Arrays;
+
+public class PrintClass {
+    public static <T, S> void printArray(T[] t, S[] s) {
+        for (T t1 : t) {
+            System.out.print(t1.toString() + " ");
+        }
+        System.out.println();
+        for (S s1 : s) {
+            System.out.println(s1.toString() + " ");
```
Comment (asecu) [inline iuriMorari/src/main/java/Generics/PrintClass.java:12]: println or print? Please test this method

**Review by asecu — COMMENTED** [review]
Take a look on the leaved comments, let me know if there are some moments to be clarified

File: iuriMorari/src/main/java/Generics/PrintClass.java (hunk 15 above)
Comment (iMarari) [reply iuriMorari/src/main/java/Generics/PrintClass.java:12]: print looks better

File: iuriMorari/src/main/java/Generics/Video.java (hunk 14 above)
Comment (iMarari) [reply iuriMorari/src/main/java/Generics/Video.java:6]: Integer can be used in collections with object reference or with generics.
int cant cause its a primitive type.
fixed

File: iuriMorari/src/main/java/Generics/PrintClass.java (hunk 12 above)
Comment (iMarari) [reply iuriMorari/src/main/java/Generics/PrintClass.java:16]: fixed

File: iuriMorari/src/main/java/Generics/AllMedia.java (hunk 1 above)
Comment (iMarari) [reply iuriMorari/src/main/java/Generics/AllMedia.java:31]: fixed

File: iuriMorari/src/main/java/Generics/Books.java (hunk 3 above)
Comment (iMarari) [reply iuriMorari/src/main/java/Generics/Books.java:14]: ageName and pagesNumber now is public

File: iuriMorari/src/main/java/Generics/GenericLibrary.java (hunk 8 above)
Comment (iMarari) [reply iuriMorari/src/main/java/Generics/GenericLibrary.java:14]: renamed

File: iuriMorari/src/main/java/Generics/GenericLibrary.java (hunk 7 above)
Comment (iMarari) [reply iuriMorari/src/main/java/Generics/GenericLibrary.java:18]: renamed

File: iuriMorari/src/main/java/Generics/GenericLibrary.java (hunk 9 above)
Comment (iMarari) [reply iuriMorari/src/main/java/Generics/GenericLibrary.java:8]: renamed

**Review by IIachimov — APPROVED** [review]

**Review by solonarieugeniu — APPROVED** [review]

**Pull request merged**